web/src/list_widget.ts

```diff
@@ -17,6 +17,31 @@ type ListWidgetMeta<Key = unknown, Item = Key> = {
     $scroll_container: JQuery;
 };
 
+type ListWidgetOpts<Key = unknown, Item = Key> = {
+    name?: string;
+    get_item: (key: Key) => Item;
+    modifier: (item: Item) => string;
+    init_sort?: string | SortingFunction<Item>;
+    initially_descending_sort?: boolean;
+    html_selector?: (item: Item) => JQuery;
+    callback_after_render?: () => void;
+    post_scroll__pre_render_callback?: () => void;
+    get_min_load_count?: (rendered_count: number, load_count: number) => number;
+    is_scroll_position_for_render?: (scroll_container: HTMLElement) => boolean;
+    filter?: {
+        $element: JQuery;
+        predicate?: (item: Item, value: string) => boolean;
+        filterer?: (list: Item[], value: string) => Item[];
```

Comment:
We can type `predicate` and `filterer` as being mutually exclusive.

Reply:
Oh right we should do that 👍